ssz-rs/src/ser.rs

```diff
@@ -55,32 +55,39 @@ pub trait Serialize {
     fn serialize(&self, buffer: &mut Vec<u8>) -> Result<usize, SerializeError>;
 }
 
+// Part represents either a fixed sized part of the serialization
+// or an offset pointing to a variably sized part of the serialization
+pub enum Part {
+    Fixed(Vec<u8>),
+    Offset(usize),
+}
+
 pub fn serialize_composite_from_components(
-    mut fixed: Vec<Option<Vec<u8>>>,
+    parts: Vec<Part>,
     mut variable: Vec<u8>,
-    variable_lengths: Vec<usize>,
     fixed_lengths_sum: usize,
+    variable_lengths_sum: usize,
     buffer: &mut Vec<u8>,
 ) -> Result<usize, SerializeError> {
-    debug_assert_eq!(fixed.len(), variable_lengths.len());
-
-    let total_size = fixed_lengths_sum + variable_lengths.iter().sum::<usize>();
+    let total_size = fixed_lengths_sum + variable_lengths_sum;
     if total_size as u64 >= MAXIMUM_LENGTH {
         return Err(SerializeError::MaximumEncodedLengthExceeded(total_size))
     }
 
     // SAFETY: `fixed_lengths_sum` fits in `u32` if the total size check holds
     let mut running_length = fixed_lengths_sum as u32;
-    debug_assert_eq!(fixed.len(), variable_lengths.len());
-    for (part_opt, variable_length) in fixed.iter_mut().zip(variable_lengths) {
-        if let Some(part) = part_opt {
-            buffer.append(part);
-        } else {
-            // SAFETY: `variable_length` fits in `u32` if the total size check holds
-            let bytes_written = running_length.serialize(buffer)?;
-            debug_assert_eq!(bytes_written, BYTES_PER_LENGTH_OFFSET);
-
-            running_length += variable_length as u32;
+    for part in parts {
+        match part {
+            Part::Fixed(mut data) => {
+                buffer.append(&mut data);
+            }
+            Part::Offset(offset) => {
+                let bytes_written = running_length.serialize(buffer)?;
+                debug_assert_eq!(bytes_written, BYTES_PER_LENGTH_OFFSET);
```

Comment:
this kind of just moves the problem elsewhere but we could change the defn of `Part`:

```rust
enum Part {
  Fixed(Vec<u8>),
  Offset(u32),
}
```